src/main/kotlin/org/jqassistant/tooling/intellij/plugin/editor/config/YamlConfigReferenceContributor.kt

```diff
@@ -0,0 +1,112 @@
+package org.jqassistant.tooling.intellij.plugin.editor.config
+
+import com.intellij.patterns.PlatformPatterns
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiLanguageInjectionHost
+import com.intellij.psi.PsiReference
+import com.intellij.psi.PsiReferenceContributor
+import com.intellij.psi.PsiReferenceProvider
+import com.intellij.psi.PsiReferenceRegistrar
+import com.intellij.psi.filters.ElementFilter
+import com.intellij.psi.filters.position.FilterPattern
+import com.intellij.psi.util.PsiTreeUtil
+import com.intellij.util.ProcessingContext
+import com.jetbrains.rd.util.forEachReversed
+import org.jetbrains.yaml.psi.YAMLKeyValue
+import org.jetbrains.yaml.psi.YAMLPsiElement
+import org.jetbrains.yaml.psi.YAMLScalar
+import org.jqassistant.tooling.intellij.plugin.data.rules.JqaRuleType
+import org.jqassistant.tooling.intellij.plugin.editor.rules.RuleReference
+
+class YamlPathFilter(
+    vararg val path: String,
+) : ElementFilter {
+    override fun isAcceptable(element: Any?, context: PsiElement?): Boolean {
+        var yamlPsiElement: YAMLPsiElement = element as? YAMLPsiElement ?: return false
+
+        path.forEachReversed { nextKey ->
+            val parent =
+                PsiTreeUtil.getParentOfType(
+                    yamlPsiElement,
+                    YAMLKeyValue::class.java,
+                ) ?: return@isAcceptable false
+
+            if (parent.keyText == nextKey) {
+                yamlPsiElement = parent
+            } else {
+                return@isAcceptable false
+            }
+        }
```

Comment:
```suggestion
        val yamlPsiElement: YAMLPsiElement = element as? YAMLPsiElement ?: return false
        // Includes self
        val parents = yamlPsiElement.parentsOfType<YAMLKeyValue>(true).toList()

        for ((key, parent) in path.reversed() zip parents) {
            if (parent.keyText != key) return false
        }
```

Please add a comment explaining what this does, it is not immediately obvious that this is traversing the tree upwards to check the path matches.

Also take a look at `yamlPsiElement.parentsOfType(true)`, I think using that could improve readability

Reply:
But it is also doing unnecessary PSI traversal isn't it?